### PR TITLE
th/ethtool-options-rh1335409-1

### PR DESCRIPTION
**Pull request by thom311**
Add libnm settings for ethtool's `--feature` / `--offload` ([rh1335409](https://bugzilla.redhat.com/show_bug.cgi?id=1335409))

This branch currently only adds the new settings to libnm. The settings are not yet used.

[2018-08-08: now, also server side is implemented. The feature is complete and ready for review]

The branch introduces a new way how NMSetting can implement properties. This is done, because the number of `--feature` options is already large (and we will add more options in the future). Also, we probably want to add support for `-C` / `--coalesce` options, so it must be very easy to add new options to libnm. The previous way based on GObject properties does not scale for that purpose.

### CODE REVIEW

File: clients/common/nm-client-utils.c [hunk 1]
```diff
@@ -132,7 +132,7 @@ nmc_string_to_bool (const char *str, gboolean *val_bool, GError **error)
 }
 
 gboolean
-nmc_string_to_tristate (const char *str, NMCTriStateValue *val, GError **error)
```
Comment (bengal) [inline clients/common/nm-client-utils.c]: The commit message says "NMTristate" instead of "NMTernary".

File: shared/nm-utils/nm-shared-utils.c [hunk 2]
```diff
@@ -1350,6 +1350,63 @@ nm_utils_strv_make_deep_copied (const char **strv)
...
+                           gboolean treat_null_as_empty,
+                           NMUtilsHashTableEqualFunc equal_func)
+{
+	guint n;
+	GHashTableIter iter;
+	gconstpointer key, v_a, v_b;
+
+	if (a == b)
+		return TRUE;
+	if (treat_null_as_empty) {
```
Comment (bengal) [inline shared/nm-utils/nm-shared-utils.c]: Shouldn't it be "if (!treat_null_as_empty)" ?

**Comment by bengal** [comment]
I'm trying to understand why this approach is better than implementing
each feature as a single GObject property.

For clients using libnm, there seem to be no advantage as anyway libnm
exposes get/set/clear feature accessor functions. A disadvantage is
that the single properties are not GObject introspectable and so
clients should use a different way to enumerate properties
(_nm_setting_gendata_get_all). But still, there is no way for example
to subscribe to property-change events for a single property, which is
probably a useful feature.

On D-Bus, I think there is no change compared to the single-GObject
properties case.

Internally, the clear advantage is that we don't need to implement
each GObject property, but this doesn't seem a major problem to me.

**Comment by thom311** [comment]
For ethtool we might add *a lot* of options. `ethtool -k` with my kernel gives about 43 features + there are 22 `--coalesce` options. So, handling these options needs to be done generically (instead of handling each one by one).
For example, there won't be an `NMSettingEthtoolPrivate.feature_rx` field. Instead, there is `NM_ETHTOOL_ID_FEATURE_RX`, which is an identifier for this feature.

---

Let's first compare this with `NMSettingBond` and `NM_SETTING_BOND_OPTIONS`. It wasn't me who implement `NMSettingBond`, but apparently adding GObject properties for each option was considered a scaling issue already previously (not by me). Granted, `NM_SETTING_BOND_OPTIONS` is considered a mistake ([rh1032808](https://bugzilla.redhat.com/show_bug.cgi?id=1032808)). But the idea of handling properties in a generic way, is still a good one. NMSettingBond however didn't strongly enough embrace the idea of an `NMSetting` being implemented as a hash. And it's implementation was a hack and it was not reusable by other implementations. So, implementing NMSettingBond was a large effort, and none of it was reusable. On top of it, in many aspects it integrates badly. The nmcli implementation was messed up (`nmcli connnection modify $CON +bond.options '...'`), and we don't have GTK doc (look at bond options in `man nm-settings`).

Of course, there is no reason why nmcli has to behave like that, and why we couldn't generate nm-settings manual with anything but GTK doc of GObject properties.

The only difference between `NM_SETTING_BOND_OPTIONS` and gendata settings is that instead of implementing the entire options handling in `nm-setting-bond.c`, it's added to the parent `nm-setting.c` (for re-use). So, it natively integrates with keyfile operations, nm_setting_to_dbus(), compare(), instead of awkwardly going through the `NM_SETTING_BOND_OPTIONS` property. For example, for `nm_setting_duplicate()` of a bond, NMSettingBond first needs to create a GHash of the values, and then set it again. That can be bypassed with gendata. I am not talking about the overhead of creating an intermediary hash. I am talking of the overhead of implementing this basic functionality of duplicating a hash in NMSettingBond and NMSettingEthtool.


Btw, note that most libnm users cannot treat properties truly generic. Meaning: they can only handle a setting property, if they know what it means (maybe nmcli is an exception, if nmcli is seen as a very dumb editor of properties. But at least, nmcli also needs to/from string functionality, which GObject does not provide either). Usually, you need to understand what properties exist and what they mean. Unknown properties are often meaningless to the libnm user, and handling them truly generic (without knowing anything about them aside the name and the type) becomes not very useful. What should however be convenient, is to treat all known properties in a similar way, look at `nm-keyfile.c` and `clients/common/nm-meta-setting-desc.c` how that is achievable. Also, `NMEthtoolID` is a way how to talk about ethtool options in a generic way.




> Internally, the clear advantage is that we don't need to implement
each GObject property, but this doesn't seem a major problem to me.

A lot of your criticism, seems about not having GObject properties for `NM_SETTING_ETHTOOL_FEATURE_RX`. gendata is an implementation detail, how NMSettingEthtool stores its values. So, while I was arguing we don't need GObject properties at all, there really is no conflict with the branch. In fact, we could add GObject properties generically like,

```C
    for (ethtool_id = _NM_ETHTOOL_ID_FEATURE_FIRST; ethtool_id <= _NM_ETHTOOL_ID_FEATURE_LAST; ethtool_id++) {
        obj_properties[ethtool_id] = g_param_spec_enum (nm_ethtool_data[ethtool_id].optname, "", "",
                                                        nm_ternary_get_type (),
                                                        NM_TERNARY_DEFAULT,
                                                        NM_SETTING_PARAM_FUZZY_IGNORE |
                                                        G_PARAM_READWRITE | 
                                                        G_PARAM_CONSTRUCT |
                                                        G_PARAM_STATIC_STRINGS);
    }
```

Any by there way, there is NM_SETTING_ETHTOOL_OPTS property. Is that one useful at all? Probably not, and maybe it should be dropped. The point is, we can add all GObject properties we want (or not). The important part is, that it should be simple (low effort) to add more features. That essentially means, to handle the properties in a truly generic way. Our way of adding GObject properties is not very generic and requires a lot of code. 


> But still, there is no way for example
> to subscribe to property-change events for a single property, which is
> probably a useful feature.

Why would you ever want to subscribe to property-changed notifications of an NMSetting? If the NMConnection/NMSetting comes from NMClient's cache, subscribing to these properties will not do you any good, because every time the profile changes (at all), the NMRemoteConnection instance replaces all NMSetting instances (and the notify signal is not fired). If you create a NMSimpleConnection instance in your application, why would you listen for notifications which are entirely triggered by your own code?






Another nice thing about gendata is, that to-dbus() and from-dbus() become lossless. That means, when libnm receives an `NMSettingEthtool` setting on D-Bus, it will contain *all* the properties. That is not the case with the GObject based implementations, which throw away the properties that are not recognized. This of course becomes most interesting, when and old libnm version talks with a newer NetworkManager version (which supports new ethtool features). In this case, `NMSettingEthtool` can contain all the features, even if it doesn't know them yet (forward compatible). Sure, `nm_setting_verify()` will then complain about them, but that is just the right thing to do. You can create an `NMSettingEthtool` setting with invalid settings, without loosing the invalid parts. But then `nm_setting_verify()` will tell you that there are some unknown/invalid fields.






TL;DR: gendata + NMSettingEthtool is just an implementation detail, for how to handle a larger number of properties, without special per-property code. We can add GObject properies as we want, but we don't actually need them for anything (provided, we add another way to generate GTK docs).

**Comment by bengal** [comment]
> A lot of your criticism, seems about not having GObject properties for NM_SETTING_ETHTOOL_FEATURE_RX. gendata is an implementation detail, how NMSettingEthtool stores its values. So, while I was arguing we don't need GObject properties at all, there really is no conflict with the branch. In fact, we could add GObject properties generically like,

Correct, plus the get_property()/set_property() accessors.

> Why would you ever want to subscribe to property-changed notifications of an NMSetting? If the NMConnection/NMSetting comes from NMClient's cache, subscribing to these properties will not do you any good, because every time the profile changes (at all), the NMRemoteConnection instance replaces all NMSetting instances (and the notify signal is not fired). If you create a NMSimpleConnection instance in your application, why would you listen for notifications which are entirely triggered by your own code?

Subscribing to property-changed is probably a bad example, but I think there is value in having GObject properties. For example nmtui binds each setting's property to a UI component and result is a very terse implementation of many setting pages.

> TL;DR: gendata + NMSettingEthtool is just an implementation detail, for how to handle a larger number of properties, without special per-property code. 

In the current branch, gendata is also a new public API of libnm to access the setting properties. Are you implying that those  nm_setting_gendata_* functions can be dropped from API?

**Comment by thom311** [comment]
> In the current branch, gendata is also a new public API of libnm to access the setting properties. Are you implying that those nm_setting_gendata_* functions can be dropped from API?

Yes, I just dropped them (before reading your comment here :) ). It *can* be public API, and it *might* be useful. But it's not necessary at this point (and maybe never will be, because it's hard to prove that something is really really necessary).

**Comment by thom311** [comment]
> Subscribing to property-changed is probably a bad example, but I think there is value in having GObject
> properties. For example nmtui binds each setting's property to a UI component and result is a very terse > implementation of many setting pages.

in this example, a complex `NM_SETTING_ETHTOOL_OPTS` property (of type `GHashTable`, or similar) wouldn't be helpful to nmtui. So, you suggest that each feature has a `NMTernary` typed property. Ok, makes sense. I could add that in a follow-up commit (though I am still not convinced that it is useful enough).

**Comment by bengal** [comment]
> in this example, a complex NM_SETTING_ETHTOOL_OPTS property (of type GHashTable, or similar) wouldn't be helpful to nmtui. So, you suggest that each feature has a NMTernary typed property. Ok, makes sense. I could add that in a follow-up commit (though I am still not convinced that it is useful enough).

As you said, we can always add GObject properties at a later time. For now, without the gendata public API, it is fine for me.

**Comment by thom311** [comment]
Now also server bits are done. Entire feature is ready for review.

File: src/settings/plugins/ifcfg-rh/nms-ifcfg-rh-writer.c [hunk 3]
```diff
@@ -809,7 +809,7 @@ write_wireless_setting (NMConnection *connection,
 	const char *device_mac, *cloned_mac;
 	guint32 mtu, chan, i;
 	gboolean adhoc = FALSE, hex_ssid = FALSE;
-	const char * const *macaddr_blacklist;
+	const char *const*macaddr_blacklist;
```
[the remark this answers is not in the document]
Comment (thom311) [reply src/settings/plugins/ifcfg-rh/nms-ifcfg-rh-writer.c]: yes, this whitespace change is silently sneaked in. I don't think it warrants a separate commit...

File: src/settings/plugins/ifcfg-rh/nms-ifcfg-rh-reader.c [hunk 4]
```diff
@@ -4046,158 +4046,143 @@ wireless_connection_from_ifcfg (const char *file,
 }
```
[the remark this answers is not in the document]
Comment (thom311) [reply src/settings/plugins/ifcfg-rh/nms-ifcfg-rh-reader.c]: Maybe. The extreme counter-example would be a rule, that no function can have more than 3 expression-statements or control-flow-statements. Otherwise would the rule require that code must be split into static functions. Clearly, that would be possible, but totally unreadable.

Likewise, I don't agree that a excessively long function is *necessarily* bad style. If there is no re-usable code, and the entire function consists of a series of steps which is performed in a logical simple manner (no complex nested loops, or complex gotos), splitting it into smaller static functions may provide little benefit. Often that applies to [parsing code](https://gitlab.gnome.org/GNOME/NetworkManager-openvpn/blob/d80d163002f936152cd3ef2230a15efa0683af18/properties/import-export.c#L776).

Whether a static function (which is called only once!) is useful, depends on the circumstance and personal preference.

[the remark this answers is not in the document]
Comment (thom311) [reply src/settings/plugins/ifcfg-rh/nms-ifcfg-rh-reader.c]: > Likewise, I don't agree

Makes it sound like I don't agree with you, although you didn't say that. I meant: "I don't agree with the common wisdom about long functions ...".

File: contrib/scripts/checkpatch.pl [hunk 5]
```diff
@@ -129,6 +129,7 @@ sub complain
 complain ("Don't use \"$1 $2\" instead of \"$2 $1\"") if $line =~ /\b(char|short|int|long) +(unsigned|signed)\b/;
 complain ("Don't use \"unsigned int\" but just use \"unsigned\"") if $line =~ /\b(unsigned) +(int)\b/;
 complain ("Please use LGPL2+ for new files") if $is_patch and $line =~ /under the terms of the GNU General Public License/;
+complain ("Don't use space inside elvis operator ?:") if $line =~ /\?[\t ]+:/;
```
Comment (lkundrak) [inline contrib/scripts/checkpatch.pl]: \s is better than [\t ]

**Comment by thom311** [comment]
Merged to master.
